lib/runtime/src/transports/etcd.rs

```diff
@@ -311,96 +314,138 @@ impl Client {
+                        let deadline = std::time::Instant::now() + Duration::from_secs(10);
+                        if let Err(e) = connector.reconnect(deadline).await {
+                            tracing::error!(
+                                "Failed to reconnect to ETCD for watching prefix '{}': {}",
+                                prefix_str,
+                                e
+                            );
                             return;
-                        };
+                        }
+                        continue;
+                    }
+                };
 
-                        // Process events
-                        for event in response.events() {
-                            // Extract the KeyValue if it exists
-                            let Some(kv) = event.kv() else {
-                                continue; // Skip events with no KV
+                // Watch the stream
+                loop {
+                    tokio::select! {
+                        maybe_resp = watch_stream.next() => {
+                            // Handle the watch response
+                            let response = match maybe_resp {
+                                Some(Ok(res)) => res,
+                                Some(Err(err)) => {
+                                    tracing::warn!(error = %err, "Error watching stream for prefix '{}'", prefix_str);
+                                    break; // Exit to reconnect
+                                }
+                                None => {
+                                    tracing::warn!("Watch stream unexpectedly closed for prefix '{}'", prefix_str);
+                                    break; // Exit to reconnect
+                                }
                             };
 
-                            // Handle based on event type
-                            match event.event_type() {
-                                etcd_client::EventType::Put => {
-                                    if let Err(err) = tx.send(WatchEvent::Put(kv.clone())).await {
-                                        tracing::error!("kv watcher error forwarding WatchEvent::Put: {err}");
-                                        return;
-                                    }
+                            // Update revision for reconnect
+                            start_revision = match response.header() {
+                                Some(header) => header.revision() + 1,
+                                None => {
+                                    tracing::error!("Missing header in watch response for prefix '{}'", prefix_str);
+                                    return;
                                 }
-                                etcd_client::EventType::Delete => {
-                                    if tx.send(WatchEvent::Delete(kv.clone())).await.is_err() {
-                                        return;
+                            };
+
```

Comment:
There's a lot of indenting in here. Do you want to try breaking out some sections to a function or method? That could come out quite tidy.